clang/lib/CIR/Dialect/Transforms/TargetLowering/CIRContext.h

```diff
@@ -0,0 +1,74 @@
+//===- CIRContext.h - Context to lower CIR ----------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// Partially mimics AST/ASTContext.h. The main difference is that this is
+// adapted to operate on the CIR dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_CIRCONTEXT_H
+#define LLVM_CLANG_LIB_CIR_DIALECT_TRANSFORMS_TARGETLOWERING_CIRCONTEXT_H
+
+#include "mlir/IR/MLIRContext.h"
+#include "mlir/IR/Types.h"
+#include "clang/AST/Type.h"
+#include "clang/Basic/TargetInfo.h"
+#include "llvm/ADT/IntrusiveRefCntPtr.h"
+
+namespace mlir {
+namespace cir {
+
+// FIXME(cir): Most of this is type-related information that should already be
+// embedded into CIR. Maybe we can move this to an MLIR interface.
```

Comment:
Agreed, and it should also supersede `CIRGenTypeCache`.

Looks like this also belongs outside of this directory, something like a `CIRContext` should be part of the top level dialect directory, and assumed to be available for use by CIRGen as well. I'm not sure I clearly see it being done as an interface, can you elaborate on that?

Reply:
> it should also supersede CIRGenTypeCache.

Not sure about this. This class has a broader scope than types as it is extracted from `ASTContext.h`

> should be part of the top level dialect directory, and assumed to be available for use by CIRGen as well.

Not sure about this either. While CIRGen uses `ASTContext.h` to do AST->LLVM queries, this aims to mimic `ASTContext.h` while adapting it to perform CIR->CIR queries. I don't think `CIRGen` will have a use for this since it already has `ASTContext.h`.

> I'm not sure I clearly see it being done as an interface, can you elaborate on that?

I'm using mostly type queries from here (e.g. `getTypeInfo`, `getTypeSize`, `getCharWidth`), this class is just abstraction to do so easily. Kinda like an interface. The idea would be to officially make this an MLIR interface. But, as mentioned above, its scope might expand beyond types.

Reply:
> I'm concerned about the CharTy below, I don't wanna see these things repeated over and over.

I agree. This `CharTy` is just used to get type information that should already be available in CIR. This is why I see an interface here: add some methods that mimic the `ASTContext.h` but gets info from CIR types instead.

I started this way to see if this class' scope would be more than types. Also to have a better idea of what an interface that replaces it would need.

> I think we need a better name then, it's misleading to name something CIRContext that isn't available or can't be used by CIRGen.

Sounds good.

Let's align what to do here then. Should I (1) move this up, (2) rename it and leave like this for now, or (3) start with the interface out of the box?

Reply:
(2) sounds good for now. You could go for (1) + (3) once you have a better grasp of how this is going to play.